### PR TITLE
Make frozen-abi optional in solana-program

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem

frozen-abi is effectively a test-only dependency, see #1096 

#### Summary of Changes
- Makes frozen-abi an optional dependency of solana-program via a new `frozen-abi` feature.
- Puts frozen-abi usage in solana-program behind `#[cfg()]` and `#[cfg_attr()]`
- Activates the `frozen-abi` feature in the dev-dependencies of solana-sdk ~and solana-program itself~

### CODE REVIEW

File: sdk/program/Cargo.toml [hunk 1]
```diff
@@ -32,8 +32,8 @@ serde_bytes = { workspace = true }
 serde_derive = { workspace = true }
 sha2 = { workspace = true }
 sha3 = { workspace = true }
-solana-frozen-abi = { workspace = true }
-solana-frozen-abi-macro = { workspace = true }
+solana-frozen-abi = { workspace = true, optional = true }
+solana-frozen-abi-macro = { workspace = true, optional = true}
```
Comment (CriesofCarrots) [inline sdk/program/Cargo.toml]:
```suggestion
solana-frozen-abi-macro = { workspace = true, optional = true }
```
I think `cargo sort` will complain about this

**Comment by joncinque** [comment]
Thanks for taking this on! The idea of featurizing frozen-abi makes a ton of sense, and I think we can take this one step further and *only* enable it when needed, which is when we run `cargo +nightly test` in CI at:

* https://github.com/anza-xyz/agave/blob/2bc026d092db26e5dfc481b68a95288784549d51/scripts/coverage.sh#L74
* https://github.com/anza-xyz/agave/blob/2bc026d092db26e5dfc481b68a95288784549d51/scripts/coverage.sh#L85
* https://github.com/anza-xyz/agave/blob/2bc026d092db26e5dfc481b68a95288784549d51/scripts/coverage-in-disk.sh#L75-L76

The work might need to go the other direction, starting with the top-level crates, adding the `frozen-abi` trait and `cfg_attr`, and working down to `solana-program`. If we start at the bottom with `solana-program`, then just doing a normal `cargo +nightly test` will fail until the top-level crates are done.

Looking through the repo, the crates that use AbiExample are:

* accounts-db
* bloom
* core
* gossip
* ledger
* perf
* program-runtime
* vote-program
* runtime
* sdk
* program
* svm
* version
* vote

I think the order would roughly go:

* core
* gossip
* ledger
* runtime
* accounts-db
* svm
* program-runtime
* perf
* vote-program
* vote
* version
* bloom
* sdk
* program

What do you think about this approach? Then only nightly tests will ever use frozen-abi, and no dapp dev will ever complain about a stack violation when building their bpf program :fire:

**Comment by kevinheavey** [comment]
@joncinque How about I just make frozen-abi a default feature of solana-program for now?

**Comment by joncinque** [comment]
> How about I just make frozen-abi a default feature of solana-program for now?

That would introduce a regression on `InstructionError` unfortunately... since making frozen-abi optional on solana-program will give the biggest immediate benefit, let's get this in as is. We can go through the other bits later.

**Review by joncinque — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**